### PR TITLE
BUG/API: make setitem-inplace preserve dtype when possible with PandasArray, IntegerArray, FloatingArray

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

xref #38896 (doesn't close).  In general `df[:, "A"] = foo` tries to operate in-place before falling back to casting.  This makes sure we do that when `foo` is a `PandasArray`, `IntegerArray`, or `FloatingArray`

I guess could/should do the same for BooleanArray

### CODE REVIEW

File: pandas/core/dtypes/missing.py [hunk 1]
```diff
@@ -506,14 +506,31 @@ def array_equals(left: ArrayLike, right: ArrayLike) -> bool:
...
+def infer_fill_value(val, length: int):
     """
     infer the fill value for the nan/NaT from the provided
     scalar/ndarray/list-like if we are a NaT, return the correct dtyped
     element to provide proper block construction
     """
     if not is_list_like(val):
         val = [val]
+
+    if type(val).__name__ == "PandasArray":
```
Comment (jbrockmendel) [inline pandas/core/dtypes/missing.py:518]: this is awful, but the relevant tests patch PandasArray._typ so that `isinstance(val, ABCPandasArray)` returns False

Comment (jreback) [reply pandas/core/dtypes/missing.py:518]: umm why can't we patch *this* routine for that test then rather than doing this?

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -506,14 +506,31 @@ def array_equals(left: ArrayLike, right: ArrayLike) -> bool:
...
+            return val
+        if length == 0:
+            return val[:0].copy()
+
+        dtype = val.dtype
+        cls = dtype.construct_array_type()
+        return cls._from_sequence([dtype._na_value], dtype=dtype).repeat(length)
+
     val = np.array(val, copy=False)
     if needs_i8_conversion(val.dtype):
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:535]: isn't this path now covered by the above? or I guess Index hits here?

**Review by jreback — CHANGES_REQUESTED** [review]
will really need to look closely. this is adding non-trivial code.

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py:535]: td64 and dt64-naive

File: pandas/core/dtypes/missing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py]: good idea, updated.  we might be able to get rid of some other kludges the same way

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -643,8 +652,8 @@ def astype(self, dtype, copy: bool = False, errors: str = "raise"):
                 )
         return newb
 
-    def _astype(self, dtype: DtypeObj, copy: bool) -> ArrayLike:
-        values = self.values
+    @staticmethod
```
Comment (jreback) [inline pandas/core/internals/blocks.py:655]: can you move to module level instead (or maybe to pandas.core.dtypes.cast); -1 on using staticmethod

File: pandas/tests/extension/test_floating.py [hunk 4]
```diff
@@ -24,12 +24,22 @@
 from pandas.tests.extension import base
 
 
-def make_data():
+def make_data(with_nas: bool = True):
+    if with_nas:
```
Comment (jreback) [inline pandas/tests/extension/test_floating.py:28]: umm can't you just parameterize over `[pd.NA, np.nan]`?
otherwise these blocks look exactly the same

Comment (jreback) [reply pandas/tests/extension/test_floating.py:28]: alternatley just
`na_value = pd.NA if with_nas else np.nan`

File: pandas/tests/extension/test_numpy.py [hunk 5]
```diff
@@ -28,6 +32,31 @@ def dtype(request):
     return PandasDtype(np.dtype(request.param))
 
 
+orig_setitem = pd.core.internals.Block.setitem
```
Comment (jreback) [inline pandas/tests/extension/test_numpy.py:35]: can you use monkeypatch instead?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:655]: sure.  and as is traditional ill point out: this is yet another kludge that wont be necessary with 2D EAs

File: pandas/tests/extension/test_numpy.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/extension/test_numpy.py:35]: this does use monkeypatch.  the monkeypatched method calls the original method

File: pandas/tests/extension/test_floating.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/extension/test_floating.py]: updated

File: pandas/core/dtypes/missing.py [hunk 6]
```diff
@@ -506,14 +506,27 @@ def array_equals(left: ArrayLike, right: ArrayLike) -> bool:
...
     infer the fill value for the nan/NaT from the provided
     scalar/ndarray/list-like if we are a NaT, return the correct dtyped
     element to provide proper block construction
     """
     if not is_list_like(val):
         val = [val]
+
+    if is_extension_array_dtype(val):
+        # We cannot use dtype._na_value bc pd.NA/pd.NaT do not preserve dtype
+        if len(val) == length:
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:520]: i am not sure how this can every be not true.

if val is an array this is always true?
if its a scalar it is as well rigth?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py:520]: updated + green

**Comment by jbrockmendel** [comment]
rebased + green

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py:520]: ill give this some thought in the AM.  this is a difficult function to describe.

**Comment by jbrockmendel** [comment]
> ok maybe if you can add a doc-string to the function will be more obvious that we are trying to match the len here

docstring updated + green

File: pandas/tests/extension/test_floating.py [hunk 7]
```diff
@@ -25,12 +25,17 @@
 from pandas.tests.extension import base
 
 
-def make_data():
+def make_data(with_nas: bool = True):
+    na_value = pd.NA if with_nas else np.nan
+    # `not with_nas` -> case without pd.NA that can be cast to
+    #  floating ndarray losslessly
+    # TODO: get cases with np.nan instead of pd.NA GH#39039
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_floating.py:32]: I don't really understand why this is needed? What does it test?

Comment (jbrockmendel) [reply pandas/tests/extension/test_floating.py:32]: Until #39039 is addressed so it is possible to make entries with np.nan instead of pd.NA, nothing.  Once that is addressed, this creates cases where it is possible to set a FloatingArray into a Series/column backed by ndarray[float64]

Comment (jorisvandenbossche) [reply pandas/tests/extension/test_floating.py:32]: Then I would leave that out until that is addressed. Because now this only doubles the test run for `test_floating.py` for nothing.

File: pandas/tests/extension/test_integer.py [hunk 8]
```diff
@@ -193,7 +197,20 @@ class TestGetitem(base.BaseGetitemTests):
 
 
 class TestSetitem(base.BaseSetitemTests):
-    pass
+    def test_setitem_series(self, data, full_indexer):
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_integer.py:200]: Can you indicate here why it is overriding the base class?

File: pandas/tests/extension/test_integer.py [hunk 9]
```diff
@@ -193,7 +197,20 @@ class TestGetitem(base.BaseGetitemTests):
...
+        ser = expected = pd.Series(data, name="data")
+        result = pd.Series(index=ser.index, dtype=object, name="data")
+
+        key = full_indexer(ser)
+        result.loc[key] = ser
+
+        if not data._mask.any():
+            # GH#38896 like we do with ndarray, we set the values inplace
+            #  but cast to the new numpy dtype
+            expected = pd.Series(data.to_numpy(data.dtype.numpy_dtype), name="data")
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_integer.py:211]: Why are we converting to the numpy dtype here? That's also not the original dtype?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Can you summarize which behaviours are changed? 
Eg from the description I would suppose that

```
df = pd.DataFrame({'a': [1, 2, 3]})
df.loc[:, 'a'] = pd.array([3, 4, 5])
```
changed behaviour to preserve the original dtype, but I don't see that directly tested?

Does this need a whatsnew?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
am ok with this, @jorisvandenbossche  if any addl comments.

**Comment by jorisvandenbossche** [comment]
There are some questions in my last review above for which I was still waiting on an answer

File: pandas/tests/extension/test_integer.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/tests/extension/test_integer.py:200]: comment added

**Comment by jbrockmendel** [comment]
let's stick a pin in this until #39163 goes in; it will be easier to address outstanding questions/comments at that point

**Comment by jbrockmendel** [comment]
mothballing until after #39163
